### PR TITLE
WIP BAN-2027 Add Apple News text, paragraph and headlines elements

### PR DESCRIPTION
**Pull request by damiendillon**
Add the
- base
- text
- heading and
- paragraph

elements, along with their spec files.

Each element is returned as a component in the Apple News format with the required keys of `role` and `text` (Later this may be expanded to contain `layout`).

Neither is the key of `format` is applied- as it is not a required field, but we the Apple News format supports HTML and we are returning HTML.

We may wish to consider adding `format` at a later point.

RE `format` the Apple docs note:
> If format is set to html or markdown, neither Additions nor
> InlineTextStyles are supported.
> Default: none
> Possible values: markdown, html, none

### CODE REVIEW

File: lib/article_json/export/apple_news/elements/paragraph.rb [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+  module Export
+    module AppleNews
+      module Elements
+        class Paragraph < Base
+          # Generate the paragraph node with its containing text elements
+          # @return [Hash]
+          def export
+            {
+              role: 'body',
+              text: text
```
Comment (damiendillon) [inline lib/article_json/export/apple_news/elements/paragraph.rb]: I considered that, as we are returning HTML, that we should possibly tag paragraphs in 'p' tags  in order to maintain consistent HTML, but also thought it may be superfluous. 


Opinions?

```suggestion
              text: "<p>#{text}</p>"
```

**Comment by damiendillon** [comment]
WIP, excuse me.

File: lib/article_json/export/apple_news/elements/paragraph.rb (hunk 1 above)
Comment (agustibr) [reply lib/article_json/export/apple_news/elements/paragraph.rb]: In my opinion, it makes sense, because of the documentation comment 

> `# Generate the paragraph node with its containing text elements`

File: lib/article_json/export/apple_news/elements/base.rb [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+    module AppleNews
+      module Elements
+        class Base
+          include ArticleJSON::Export::Common::Elements::Base
+
+          # Export the given element. Dynamically looks up the right
+          # export-element-class, instantiates it and then calls the `#export`
+          # method.
+          # Defaults to an empty string, e.g. if no exporter is specified for
+          # the given type.
```
Comment (dsager) [inline lib/article_json/export/apple_news/elements/base.rb]: The default is `nil`, isn't it?

File: lib/article_json/export/apple_news/elements/paragraph.rb (hunk 1 above)
Comment (dsager) [reply lib/article_json/export/apple_news/elements/paragraph.rb]: This depends on how Apple defines the format. From the example [here](https://developer.apple.com/documentation/apple_news/body#2973276) I understand that it's not necessary and that `{"role":  "body","text": "some text of the paragraph"}` is a valid paragraph.

Paragraph is what we call this type of element, Apple calls it body. So I don't think it needs to include a `p` tag...

File: lib/article_json/export/apple_news/elements/text.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
+module ArticleJSON
+  module Export
+    module AppleNews
+      module Elements
+        class Text < Base
+          include ArticleJSON::Export::Common::HTML::Elements::Base
+          include ArticleJSON::Export::Common::HTML::Elements::Text
+
+          def export
```
Comment (dsager) [inline lib/article_json/export/apple_news/elements/text.rb]: I think it'd be worth adding a bit of documentation here, why are we using super and calling to_s on it?

File: lib/article_json/import/google_doc/html/image_parser.rb [hunk 4]
```diff
@@ -22,25 +22,31 @@ def initialize(node:, caption_node:, css_analyzer:)
           # The value of the image's `alt` attribute
           # @return [String]
           def alt
+            return '' if image_url?
```
Comment (dsager) [inline lib/article_json/import/google_doc/html/image_parser.rb]: I think the changes in this file shouldn't be part of your PR?

File: lib/article_json/import/google_doc/html/node_analyzer.rb [hunk 5]
```diff
@@ -88,7 +88,17 @@ def quote?
           # @return [Boolean]
           def image?
             return @is_image if defined? @is_image
-            @is_image = node.xpath('.//img').length > 0
+            @is_image = image_url? || node.xpath('.//img').length > 0
```
Comment (dsager) [inline lib/article_json/import/google_doc/html/node_analyzer.rb]: Changes in this file are also not related to Apple News

File: spec/article_json/export/apple_news/elements/base_spec.rb [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    context 'when the source element is a heading' do
+      let(:source_element) do
+        ArticleJSON::Elements::Heading.new(content: 'Foo Bar', level: 1)
+      end
+      let(:heading) { { role: 'heading1', text: 'Foo Bar' } }
+      it { should eq heading }
+    end
+
+    context 'when the source element is a paragraph' do
+      let(:source_element) { sample_paragraph }
```
Comment (dsager) [inline spec/article_json/export/apple_news/elements/base_spec.rb]: I think the `sample_paragraph` above can be removed if you specify it directly in this context:
```suggestion
      let(:source_element) { ArticleJSON::Elements::Paragraph.new(content: [sample_text]) }
```

File: spec/article_json/export/apple_news/elements/heading_spec.rb [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+      it { should eq heading }
+    end
+
+    context 'when the heading level is 2' do
+      let(:level) { 2 }
+      let(:heading) { { role: 'heading2', text: 'Foo Bar' } }
+      it { should eq heading }
+    end
+
+    (3..6).each do |i|
```
Comment (dsager) [inline spec/article_json/export/apple_news/elements/heading_spec.rb]: Why not use the loop to cover all levels?
```suggestion
    (1..6).each do |i|
```

File: spec/article_json/import/google_doc/html/image_parser_spec.rb [hunk 8]
```diff
@@ -16,14 +16,19 @@
     ArticleJSON::Import::GoogleDoc::HTML::CSSAnalyzer.new(css)
```
Comment (dsager) [inline spec/article_json/import/google_doc/html/image_parser_spec.rb]: Changes in this file not related to Apple News...

File: spec/article_json/import/google_doc/html/node_analyzer_spec.rb [hunk 9]
```diff
@@ -178,12 +178,58 @@
       it { should be true }
```
Comment (dsager) [inline spec/article_json/import/google_doc/html/node_analyzer_spec.rb]: Changes in this file not related to Apple News...

File: lib/article_json/import/google_doc/html/image_parser.rb (hunk 4 above)
Comment (damiendillon) [reply lib/article_json/import/google_doc/html/image_parser.rb]: That is correct. This is part the git issue I said I had.

File: lib/article_json/import/google_doc/html/node_analyzer.rb (hunk 5 above)
Comment (damiendillon) [reply lib/article_json/import/google_doc/html/node_analyzer.rb]: That is correct. This is part the git issue I said I had.

File: spec/article_json/export/apple_news/elements/heading_spec.rb (hunk 7 above)
Comment (damiendillon) [reply spec/article_json/export/apple_news/elements/heading_spec.rb]: Good point

File: spec/article_json/import/google_doc/html/node_analyzer_spec.rb (hunk 9 above)
Comment (damiendillon) [reply spec/article_json/import/google_doc/html/node_analyzer_spec.rb]: That is correct. This is part the git issue I said I had.

File: spec/article_json/import/google_doc/html/image_parser_spec.rb (hunk 8 above)
Comment (damiendillon) [reply spec/article_json/import/google_doc/html/image_parser_spec.rb]: That is correct. This is part the git issue I said I had.
